src/billing_handler.rs

```diff
@@ -2,32 +2,73 @@ use crate::model::AppState;
 
 use actix_web::web::{Data, Json};
 use actix_web::{get, post, HttpResponse, Responder};
-use anyhow::Context;
+use k256::ecdsa::SigningKey;
 use k256::elliptic_curve::generic_array::sequence::Lengthen;
 use serde_json::json;
 use tiny_keccak::{Hasher, Keccak};
 
 #[derive(Debug, serde::Deserialize)]
 pub struct SigningData {
     nonce: String,
-    txhashes: Vec<String>,
+    tx_hashes: Vec<String>,
 }
 
 #[get("/billing/inspect")]
-pub async fn get_bill(appstate: Data<AppState>) -> impl Responder {
+pub async fn inspect_bill(appstate: Data<AppState>) -> impl Responder {
     HttpResponse::Ok().json(json!({
         "bill": appstate.execution_costs.lock().unwrap().clone(),
     }))
 }
 
+#[get("/billing/latest")]
+pub async fn get_last_bill_claim(appstate: Data<AppState>) -> impl Responder {
+    let mut last_bill_claim_guard = appstate.last_bill_claim.lock().unwrap();
+    if last_bill_claim_guard.0.is_none() {
+        return HttpResponse::Ok().body("No bill claimed yet");
+    }
+
+    if last_bill_claim_guard.1.is_some() {
+        return HttpResponse::Ok().json(json!({
+            "bill_claim_data": last_bill_claim_guard.0.clone().unwrap(),
+            "signature": last_bill_claim_guard.1.clone().unwrap(),
+        }));
+    }
+
+    let bill_claim_data = hex::decode(last_bill_claim_guard.0.clone().unwrap());
+    if bill_claim_data.is_err() {
+        return HttpResponse::InternalServerError().body(format!(
+            "Failed to decode claimed bill data: {}",
+            bill_claim_data.unwrap_err()
+        ));
+    }
+
+    let bill_claim_data = bill_claim_data.unwrap();
+    let signature = sign_data(bill_claim_data.as_slice(), &appstate.signer).await;
+    if signature.is_err() {
+        return HttpResponse::InternalServerError().body(format!(
+            "Failed to sign billing data: {}",
+            signature.unwrap_err()
+        ));
+    }
+
+    let bill_claim_data = hex::encode(bill_claim_data.as_slice());
+    let signature = signature.unwrap();
+    last_bill_claim_guard.1 = Some(signature.clone());
+
+    HttpResponse::Ok().json(json!({
+        "bill_claim_data": bill_claim_data,
+        "signature": signature,
+    }))
+}
+
 #[post("/billing/export")]
-pub async fn sign_data(appstate: Data<AppState>, data: Json<SigningData>) -> impl Responder {
+pub async fn export_bill(appstate: Data<AppState>, data: Json<SigningData>) -> impl Responder {
     let signing_data = data.into_inner();
     if signing_data.nonce.is_empty() {
         return HttpResponse::BadRequest().body("Nonce must not be empty");
     }
 
```

Comment:
Add nonce check, as suggested by @roshanr95  in other comments. Should be diff from previous nonce.

Reply:
Cool